### ISSUE TITLE
Fix ClickHouse Backup Failures

### ISSUE THREAD

**Issue by rzore** [rzore/Test#65]
We need to figure out these errors in ClickHouse, which are caused by backups.

```
2017.11.08 17:48:39.077379 [ 1012426 ] <Error> executeQuery: Code: 210, e.displayText() = DB::NetException: Connection reset by peer while writing to socket ([::1]:57574), e.what() = DB::NetException (from [::1]:57574) (in query: SELECT * FROM analytics.messagelog WHERE eventday >= '2017-10-30' AND eventday < '2017-10-31'), Stack trace:

0. clickhouse-server(StackTrace::StackTrace()+0x16) [0x3358e46]
1. clickhouse-server(DB::Exception::Exception(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, int)+0x1f) [0x14bc4df]
2. clickhouse-server(DB::WriteBufferFromPocoSocket::nextImpl()+0x3c8) [0x33826a8]
3. clickhouse-server(DB::CompressedWriteBuffer::nextImpl()+0x3dd) [0x33959dd]
4. clickhouse-server(DB::DataTypeString::serializeBinaryBulk(DB::IColumn const&, DB::WriteBuffer&, unsigned long, unsigned long) const+0x285) [0x2cdfbf5]
5. clickhouse-server(DB::NativeBlockOutputStream::writeData(DB::IDataType const&, std::shared_ptr<DB::IColumn> const&, DB::WriteBuffer&, unsigned long, unsigned long)+0x264) [0x2cbb074]
6. clickhouse-server(DB::NativeBlockOutputStream::writeData(DB::IDataType const&, std::shared_ptr<DB::IColumn> const&, DB::WriteBuffer&, unsigned long, unsigned long)+0x315) [0x2cbb125]
7. clickhouse-server(DB::NativeBlockOutputStream::write(DB::Block const&)+0xd60) [0x2cbbec0]
8. clickhouse-server(DB::TCPHandler::sendData(DB::Block&)+0x45) [0x14c7d85]
9. clickhouse-server(DB::TCPHandler::processOrdinaryQuery()+0x42f) [0x14c8d1f]
10. clickhouse-server(DB::TCPHandler::runImpl()+0x73f) [0x14c9e6f]
11. clickhouse-server(DB::TCPHandler::run()+0x2b) [0x14ca8bb]
12. clickhouse-server(Poco::Net::TCPServerConnection::start()+0xf) [0x3d2cbdf]
13. clickhouse-server(Poco::Net::TCPServerDispatcher::run()+0x13b) [0x3d3301b]
14. clickhouse-server(Poco::PooledThread::run()+0xb7) [0x3f9de87]
15. clickhouse-server(Poco::ThreadImpl::runnableEntry(void*)+0xa5) [0x3f69ec5]
16. /lib/x86_64-linux-gnu/libpthread.so.0(+0x76ba) [0x7ff3d20406ba]
17. /lib/x86_64-linux-gnu/libc.so.6(clone+0x6d) [0x7ff3d16613dd]

2017.11.08 17:48:39.100248 [ 1012426 ] <Warning> TCPHandler: Client has gone away.
```

**Comment by rzore** [comment]
➤ Matt Barton commented:

I feel like ClickHouse leaves a lot to be desired when it comes to backing up data. There aren't any apparent tools or anything simple about it.

Data Dump Strategy:

We're currently dumping data on a per-day basis, meaning that we dump all the data for a particular day and archive it. The problem with this approach is that even daily dumps can be too much data for ClickHouse to return, resulting in errors. It might be possible to switch this to being per-hour, but I don't see that as any kind of long-term solution. What happens when 1 hour's worth of data is too much?

Freeze Partitions Strategy:

The ClickHouse documentation talks about freezing partitions. Freezing partitions are effectively a filesystem snapshot, which means that it creates hard links back to the original files in a shadow directory. One can then backup those files, and then delete them after.

The problem with this approach is that freezing partitions can only be done at a per-table basis. There is no way to tell ClickHouse to freeze all partitions in all tables in all databases at once. Instead, we have to iterate through each database and their tables, issuing a ALTER TABLE ... FREEZE PARTITION command for each one. Meh!

I've got a script which can do this, but the bigger issue is that it takes a long time to archive the data. I need to do a comparison against the other backup method to be sure, but it seems like it took about four hours or so, but I don't have precise timing.

The other thing that sucks about this method is that you have to do this on ALL servers. It can be done on the replicas, if you want, but basically if we have our data spread out across four primary systems (non-replicas) then we need to run this on those four systems. I am concerned about the performance impact on the Synology with having four systems dump backup data to that box.

Next steps are to execute the backup script on all systems, and measure how long it takes for it to complete. I need to add some things to the script to help us measure the amount of time it takes.

**Comment by rzore** [comment]
➤ Matt Barton commented:

The best course of action, at this time, is to implement a script I've been working on which will automate the freezing of partitions on each server (i.e., shard), and then backup the entire /data/clickhouse/shadow into a tarball that'll be stored in /backup/clickhouse/prod/partitions (Synology). The backups will be stored on a per-server basis.

I still need to implement one important feature: to delete the contents of the /data/clickhouse/shadow directory when the backup is complete. Doingthis is SAFE and mentioned in the documentation.

**Comment by rzore** [comment]
➤ Ryan Pfenninger commented:

you probably already thought about this, but we only need to back up one set of servers.  i would back up the A's since all the writes come into the B's.

**Comment by rzore** [comment]
➤ Matt Barton commented:

The cron job to run the clickhouse_snapshot.sh script has been scheduled for 1:00 AM UTC every day on click[1234]a. If this works over the weekend, the plan will be to disable the old cron jobs and backup scripts, officially retiring them.

I think I would like to make clickhouse_snapshot.sh a weekly/full backup job, and then create some kind of incremental backup for partitions in either the current year or month.

**Comment by rzore** [comment]
➤ Matt Barton commented:

As far as I can tell, the backing up of partitions seems to work well. I forgot that the script I wrote to backup data to S3 would cover backing up the partition data files, so that's already done and they're all up there.

**Comment by rzore** [comment]
➤ Matt Barton commented:

Last night (12/18), we had problems with the ClickHouse backups going to the Synology (backup1). The Synology was overwhelmed. Nagios had a difficult time with service checks that talk to the Nagios agent, running on individual servers. The Nagios agent was getting stuck trying to pull storage stats about the /backup NFS mount, and that caused a bunch of alerts due to timeouts.

The root cause was that we overwhelmed the Synology due to two backup scripts running simultaneously. I missed that the script which uploads everything in /backup/clickhouse/prod on click1a to AWS S3 executes only one hour after clickhouse_snapshot.sh runs, which is NOT enough time. This is totally my fault.

The backup script remains unchanged from last night; however, I disabled the cron job which syncs files up to S3. As such, performance seems to be better tonight and Nagios hasn't gotten behind. No timeouts or anything, which is perfect.

Lastly, I've created #71 to address some configuration changes, updates, etc. on the Synology and the server which rely on it.
